pkg/charts/charts.go

```diff
@@ -0,0 +1,98 @@
+package charts
+
+import (
+	"bytes"
+	"path"
+
+	"github.com/linkerd/linkerd2/pkg/charts/static"
+	"k8s.io/helm/pkg/chartutil"
+	helmChart "k8s.io/helm/pkg/proto/hapi/chart"
+	"k8s.io/helm/pkg/renderutil"
+	"k8s.io/helm/pkg/timeconv"
+)
+
+// Chart holds the necessary info to render a Helm chart
+type Chart struct {
+	Name      string
+	Dir       string
+	Namespace string
+	RawValues []byte
+	Files     []*chartutil.BufferedFile
+}
+
+// Render returns a bytes buffer with the result of rendering a Helm chart
+func (chart *Chart) Render() (bytes.Buffer, error) {
+	if err := filesReader(chart.Dir+"/", chart.Files); err != nil {
+		return bytes.Buffer{}, err
+	}
+
+	// Keep this slice synced with the contents of /charts/partials
+	partialsFiles := []*chartutil.BufferedFile{
+		{Name: "charts/partials/" + chartutil.ChartfileName},
+		{Name: "charts/partials/templates/_proxy.tpl"},
+		{Name: "charts/partials/templates/_proxy-init.tpl"},
+		{Name: "charts/partials/templates/_volumes.tpl"},
+		{Name: "charts/partials/templates/_resources.tpl"},
+		{Name: "charts/partials/templates/_metadata.tpl"},
+		{Name: "charts/partials/templates/_helpers.tpl"},
+		{Name: "charts/partials/templates/_debug.tpl"},
+		{Name: "charts/partials/templates/_capabilities.tpl"},
+	}
+	if err := filesReader("", partialsFiles); err != nil {
+		return bytes.Buffer{}, nil
+	}
+
+	// Create chart and render templates
+	chrt, err := chartutil.LoadFiles(append(chart.Files, partialsFiles...))
+	if err != nil {
+		return bytes.Buffer{}, err
+	}
+
+	renderOpts := renderutil.Options{
+		ReleaseOptions: chartutil.ReleaseOptions{
+			Name:      chart.Name,
+			IsInstall: true,
+			IsUpgrade: false,
+			Time:      timeconv.Now(),
+			Namespace: chart.Namespace,
+		},
+		KubeVersion: "",
+	}
+
+	chrtConfig := &helmChart.Config{Raw: string(chart.RawValues), Values: map[string]*helmChart.Value{}}
+	renderedTemplates, err := renderutil.Render(chrt, chrtConfig, renderOpts)
+	if err != nil {
+		return bytes.Buffer{}, err
+	}
+
+	// Merge templates and inject
+	var buf bytes.Buffer
+	for _, tmpl := range chart.Files {
+		t := path.Join(renderOpts.ReleaseOptions.Name, tmpl.Name)
+		if _, err := buf.WriteString(renderedTemplates[t]); err != nil {
+			return bytes.Buffer{}, err
+		}
+	}
+
+	return buf, nil
+}
+
+func filesReader(dir string, files []*chartutil.BufferedFile) error {
+	for _, f := range files {
+		filename := dir + f.Name
+		if dir == "" {
+			filename = filename[7:]
```

Comment:
I'm not sure I understood your suggestion about using the full paths...
The paths in lines 30-39 need to remain as is for the Helm lib to understand they're dependencies. `filesReader` is not to be used outside this file, but I do agree using the empty string for logic is not very kosher. I'll refactor this to be more explicit.

Reply:
Thanks for the explanation via _slack_. Both https://github.com/linkerd/linkerd2/pull/3195#pullrequestreview-270912190 and https://github.com/linkerd/linkerd2/pull/3195/commits/41c64f1e5d11854f79b303ab507985f8af03d2be also provide the additional context. Essentially, there are two things happening here:

1. The code defines the `partials` chart as a dependency of the `linkerd2` and `patch` charts by specifying the file path prefix `charts/partials` as seen in [line 30-39](https://github.com/linkerd/linkerd2/blob/alpeb/helm-injection3/pkg/charts/charts.go#L30-L40).  (I confused the `charts` subfolder here with the `github.com/linkerd/linkerd2/charts` root folder.)
2. The `filesReader()` function reads the content of the `partials` chart directly from the `github.com/linkerd/linkerd2/charts` root folder, instead of from the git-ignored `partials.tgz` dependency file.

Reply:
I'm hoping to add the `bin/helm.sh` script to the `bin/docker_build` and `bin/go-run` scripts as part of the upcoming `install` task, to avoid errors from sneaking into the Helm charts unnoticed. (I think it's reasonable since `install`, `upgrade` and `inject` code rely on the Helm charts being correct.)

Maybe then we can try load the `charts/partials.tgz` instead of multiple YAML template files. (The `.tgz` file will always be created by the build scripts anyway.) That can help to avoid the code-and-templates out-of-sync issue as you mentioned [here](https://github.com/linkerd/linkerd2/blob/alpeb/helm-injection3/pkg/charts/charts.go#L29) and the `dir == ""` check if my first comment. WDYT?

Reply:
I'm ok with that, but  we have to be clear that that means everybody should have the `helm` binary installed, even if they won't be working on template modifications. Perhaps `bin/helm-sh` can check for the presence of that binary, check its version, and install/upgraded it if necessary. The build environments should also have it. @siggy wanted to be in the loop about that kind of build environment change, so I'd love to get his opinion as well.

Reply:
Yup agree about being super intentional around adding helm as a dev/build dependency. Are there any approaches that do not add helm as a dependency that have reasonable tradeoffs?

If not, I think as @alpeb mentioned, it's important the `bin/helm.sh` script handle downloading/installing the correct helm binary, similar to what `bin/lint` and `bin/protoc` do today. We want to ensure that the build is deterministic regardless of whether someone has helm installed, or what version they have.

Reply:
Updated https://github.com/linkerd/linkerd2/issues/3143 to track the work to incorporate Helm into our dev workflow.